src/backend/distributed/commands/foreign_server.c

```diff
@@ -0,0 +1,378 @@
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+	if (!ShouldPropagateObject(&address))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *sql = DeparseTreeNode(node);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PreprocessRenameForeignServerStmt is called during the planning phase for
+ * ALTER SERVER RENAME.
+ */
+List *
+PreprocessRenameForeignServerStmt(Node *node, const char *queryString,
+								  ProcessUtilityContext processUtilityContext)
+{
+	RenameStmt *stmt = castNode(RenameStmt, node);
+	Assert(stmt->renameType == OBJECT_FOREIGN_SERVER);
+
+	char *serverName = strVal(stmt->object);
+	ForeignServer *server = GetForeignServerByName(serverName, false);
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+	/* filter distributed servers */
+	if (!ShouldPropagateObject(&address))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *sql = DeparseTreeNode(node);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PreprocessAlterForeignServerOwnerStmt is called during the planning phase for
+ * ALTER SERVER .. OWNER TO.
+ */
+List *
+PreprocessAlterForeignServerOwnerStmt(Node *node, const char *queryString,
+									  ProcessUtilityContext processUtilityContext)
+{
+	AlterOwnerStmt *stmt = castNode(AlterOwnerStmt, node);
+	Assert(stmt->objectType == OBJECT_FOREIGN_SERVER);
+
+	char *serverName = strVal(stmt->object);
+	ForeignServer *server = GetForeignServerByName(serverName, false);
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, ForeignServerRelationId, server->serverid);
+
+	/* filter distributed servers */
+	if (!ShouldPropagateObject(&address))
+	{
+		return NIL;
+	}
+
+	EnsureCoordinator();
+
+	char *sql = DeparseTreeNode(node);
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	List *commands = list_make3(DISABLE_DDL_PROPAGATION,
+								(void *) sql,
+								ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
+}
+
+
+/*
+ * PreprocessDropForeignServerStmt is called during the planning phase for
+ * DROP SERVER.
+ */
+List *
+PreprocessDropForeignServerStmt(Node *node, const char *queryString,
+								ProcessUtilityContext processUtilityContext)
+{
+	DropStmt *stmt = castNode(DropStmt, node);
+	Assert(stmt->removeType == OBJECT_FOREIGN_SERVER);
+
+	bool includesDistributedServer = NameListHasDistributedServer(stmt->objects);
+
+	if (!includesDistributedServer)
+	{
+		return NIL;
+	}
+
+	if (list_length(stmt->objects) > 1)
```

Comment:
can you add a test to cover this line? We don't seem to have a test that properly checks the return value of `NameListHasDistributedServer`

Reply:
Added, thanks